### ISSUE TITLE
Source trait selector does not seem to be working

### ISSUE THREAD

**Issue by BugwoodCenter** [gimmefreshdata/freshdata#75]
![image](https://user-images.githubusercontent.com/32331247/30986049-3df6df08-a460-11e7-9f96-d73d82987b9d.png)
I've tried via the API and the interface to use source as a trait - it continues to give a response of either busy or 400 (bad input).  

http://gimmefreshdata.github.io/?limit=20&taxonSelector=Aves%2CInsecta&traitSelector=source%20%3D%3D%20inaturalist%20string&wktString=POLYGON%20%28%28-72.147216796875%2041.492120839687786%2C%20-72.147216796875%2043.11702412135048%2C%20-69.949951171875%2043.11702412135048%2C%20-69.949951171875%2041.492120839687786%2C%20-72.147216796875%2041.492120839687786%29%29

**Comment by jhammock** [comment]
hmm... "busy" does not indicate an error, just that the system is unavailable (that "we can handle one query at a time" constraint I mentioned earlier). 

@diatomsRcool, are you running national checklist queries currently? If so, do they run around the clock? We might need to arrange some turn-taking, as there are two users of the system.

**Comment by jhammock** [comment]
oh- but I think that should just be "source=inaturalist", where you have "source=inaturalist string"

**Comment by BugwoodCenter** [comment]
The "string" was added by the page.  It was also specified in the wiki - https://github.com/gimmefreshdata/freshdata/wiki/api
![image](https://user-images.githubusercontent.com/32331247/30987050-9fbacd28-a463-11e7-9971-19bc80ff59ce.png)

**Comment by jhammock** [comment]
I stand corrected! I haven't used that filter in ages.

**Comment by BugwoodCenter** [comment]
It's all good! :)

**Comment by diatomsRcool** [comment]
I have found that if its busy, then you just have to try again later. It 
can only handle two queries at once and I've been hitting it pretty 
hard. I also get the "bad request" when its busy. So, maybe try again later?

**Comment by jhammock** [comment]
@diatomsRcool are you launching your requests automatically, or do you have down time when the service won't be occupied (at least with your queries)?

**Comment by diatomsRcool** [comment]
I am not launching automatically, just whenever I have time and I think 
of it

**Comment by jhpoelen** [comment]
I just had a peek inside the cluster of computers that does the number crunching for freshdata / effechecka and found that @godfoder and @gauravyeole have been reserving 96 + 16 = 112 cpus and 28.5 GB of memory with jupyter notebooks that haven't been active a week or longer. (see attached screenshot). @mjcollin is running some stuff also, but this appears to be active and recent.

Because of this, the queries you are requesting cannot be handled due to lack of resources.

@godfoder @gauravyeole - could you please stop your jupyter notebooks on idigbio ?

![screenshot from 2017-09-29 09-27-25](https://user-images.githubusercontent.com/1084872/31025865-1931240a-a4f9-11e7-86d5-e41f8bad1ace.png)

I think it would be nice to have a way to alert folks of inactive python notebooks . . . perhaps make a public list of all active users, number of active jobs and the cpu/memory they are consuming?

Related to issues https://github.com/bio-guoda/guoda-services/issues/14 https://github.com/bio-guoda/guoda-services/issues/19 https://github.com/bio-guoda/guoda-services/issues/20 .

**Comment by BugwoodCenter** [comment]
I'm also happy to be put on a schedule that will send data to an endpoint when the system is idle (if that time exists).  My intent is to consume the tsv and have [AWS's Data Pipeline](https://aws.amazon.com/datapipeline/) do the heavy lifting.  It's event driven - so when the file shows up, it goes to work.  I'm not picky about when.

**Comment by diatomsRcool** [comment]
I can stay off today and next week, if you want to use it exclusively

**Comment by BugwoodCenter** [comment]
When using the test page at http://gimmefreshdata.github.io, what does it mean when you get no results back?  I submitted a test of Insecta with a bounding box around Michigan.  The first time is came back with "Submitted" and no results.  Sent the request a second time and got busy.  I know there is no que - so I was looking for some indicator of what was running.  Am I missing something fairly obvious?

**Comment by jhammock** [comment]
at this point it should be "busy" with your query. Getting the "submitted" response once is your final notice that the query should complete in due course. Has it taken more than ~20 minutes? Mine tend to come up with results after about 15

**Comment by BugwoodCenter** [comment]
It's been about 13 at this point - I'll poke it in a bit and let you know.

**Comment by jhpoelen** [comment]
could you please share your query by pasting the url into this issue?

**Comment by BugwoodCenter** [comment]
http://gimmefreshdata.github.io/?limit=20&taxonSelector=Insecta&traitSelector=&wktString=POLYGON%20%28%28-87.528076171875%2041.343824581185686%2C%20-87.528076171875%2045.89000815866184%2C%20-81.925048828125%2045.89000815866184%2C%20-81.925048828125%2041.343824581185686%2C%20-87.528076171875%2041.343824581185686%29%29

**Comment by jhpoelen** [comment]
Thanks! Just had a peek and . . . the request is being handled, but other processes are taking up most of the resources. They appear to the pyspark-shell cmds that may or may not be active but are taking up precious resources. Lack of resources slows down the freshdata queries (see https://github.com/bio-guoda/guoda-services/issues/14 for known issue). Perhaps @mjcollin or @godfoder know more about this. I've attached a screenshot of the computation framework activities for those interested. 

Thanks for your patience - I can see two things to improve (1) an indication of progress beyond "busy" and (2) make the queries faster .

![screenshot from 2017-10-12 13-28-14](https://user-images.githubusercontent.com/1084872/31517678-80b42fa2-af51-11e7-94fd-e557f716ef4f.png)

**Comment by BugwoodCenter** [comment]
Knowing that busy refers specifically to my query and not just a general state of the server is a big help.  It will keep impatient button pushers like me from adding load with incessant refreshes and alternate attempts with other parameters to see if the request was bad to start with.

**Comment by BugwoodCenter** [comment]
And I did just get a response with data.  TSV downloaded like a dream.

**Comment by jhammock** [comment]
So FTR, at the present system load this query took about an hour?

**Comment by BugwoodCenter** [comment]
Approximately.  I was not button mashing to see when it completed - so I can only say no more than an hour.  

I do not mind waiting on data - but if we can add a web hook parameter to the initial request or allow people to register a web hook that would notify that endpoint that the data is ready, it may be useful.   

I'll be a little more fastidious on recording when future requests go in and a response comes back.

**Comment by jhpoelen** [comment]
Checking for the data is cheap, so please keep pushing that button manually or using an automated script to poll once in a while (perhaps once every 5 minutes?) until the results are available. 

And yes, this time the query took about an hour, however, this was due to a bit of a traffic jam on the cluster due to open spark-shells. If this were not the case, the process would have completed faster.

**Comment by jhpoelen** [comment]
For instance, for the last example, you can create a little script that polls http://api.effechecka.org/occurrences?limit=20&taxonSelector=Insecta&traitSelector=&wktString=POLYGON%20%28%28-87.528076171875%2041.343824581185686%2C%20-87.528076171875%2045.89000815866184%2C%20-81.925048828125%2045.89000815866184%2C%20-81.925048828125%2041.343824581185686%2C%20-87.528076171875%2041.343824581185686%29%29 until the "ready" status is found.

**Comment by BugwoodCenter** [comment]
No complaints here.  5 minutes sounds reasonable for a poke.  

After initial creation of a dataset, how frequently is it updated?  I'm assuming that you'd rather someone hit an existing monitor than making a new monitor that includes a start date.

**Comment by jhpoelen** [comment]
You can poll an existing monitor for new data by using a url including the addedAfter query parameter. For previous example, you can make a list of all records added after 2019-01-01 by using something like- 

http://api.effechecka.org/occurrences?limit=20&wktString=POLYGON%20%28%28-87.528076171875%2041.343824581185686%2C%20-87.528076171875%2045.89000815866184%2C%20-81.925048828125%2045.89000815866184%2C%20-81.925048828125%2041.343824581185686%2C%20-87.528076171875%2041.343824581185686%29%29&addedAfter=2019-01-01 . 

The addedAfter and addedBefore search use an index on the existing monitor, so no need to wait. This is also a relative cheap operation, so polling should be ok.

**Comment by BugwoodCenter** [comment]
And that can be used with the UUID for the monitor - Excellent!

**Comment by jhammock** [comment]
FYI, we're still working on our schedule of resource updates, which will influence how often you are likely to get new records. I'm hoping for weekly or better for iNaturalist, less often (monthly?) for the behemoth resources.

**Comment by jhpoelen** [comment]
@BugwoodCenter @jhammock if this issue has been addressed, please close. Otherwise, please describe what is left to do.
